### ISSUE TITLE
ambient: allow more control over mesh enablement

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#53578]
In sidecars, and Ambient pre-alpha, we have quite a bit of control on what triggers something to be 'in the mesh'.

When ambient went to alpha we streamlined this to only allow a single hardcoded label opt-in by the user on the namespace.

We should consider adding more flexibility.

Use cases:
* I want to opt-in the entire cluster. This is the cluster-wide default networking stack, so make everything use it.
  * But maybe opt out a few namespace as an admin (kube-system, etc)
  * But maybe allow namespace admins to opt out (if they cannot use ambient for various reasons
* I want to _enforce_ all pods are in the mesh for security purposes

Note that while the last one is not likely something we can commit to at this time, it would be good if the API we decide is future proofed to enable that use case if/when we can guarantee it.

In the past, which was likely a solid API, we allowed a matchExpression in meshConfig to be defined. We can default, of course, to `istio.io/dataplane-mode=ambient`, but could express an opt-out model (like `istio.io/dataplane-mode NOTIN [disabled]`).

**Comment by MorrisLaw** [comment]
+1, Microsoft would be interested in the added flexibility as well.

**Comment by bleggett** [comment]
We already effectively support this with the current `istio.io/dataplane-mode` labels - `dataplane-mode: none` is supported, and can be applied to pods.

**Comment by bleggett** [comment]
Also worth noting that if dynamic enable/disable without pod restart via labels is not something people find terribly useful, we could always drop `istio-cni` and replace it with an NRI plugin or something.

**Comment by ilrudie** [comment]
> Also worth noting that if dynamic enable/disable without pod restart via labels is not something people find terribly useful, we could potentially drop `istio-cni` and do something with NRI where the pod could be injected on recycle.

I think dynamic enable without needing to bounce pods is a good selling point

**Comment by bleggett** [comment]
> > Also worth noting that if dynamic enable/disable without pod restart via labels is not something people find terribly useful, we could potentially drop `istio-cni` and do something with NRI where the pod could be injected on recycle.
> 
> I think dynamic enable without needing to bounce pods is a good selling point

I agree. I'm just throwing it in the pile for consideration, as long as we're (re)evaluating UX here.

**Comment by bleggett** [comment]
> I want to enforce all pods are in the mesh for security purposes
> Note that while the last one is not likely something we can commit to at this time, it would be good if the API we decide is future proofed to enable that use case if/when we can guarantee it.

I'm not sure this can be guaranteed [given how it works today](https://istio.io/latest/docs/ambient/usage/add-workloads/#pod-selection-logic-for-ambient-and-sidecar-modes) - at least, not without the use of a webhook or some external validator to restrict pod/ns labels (which will always work), or completely redefining label precedence (which seems like a bad idea that will confuse people)

We can of course make the autoenroll mechanism completely ignore labels and their existing precedence, but this presents a problem - 

what if I want to 

- auto-enroll everything
- but I want to exclude `istio-system` or `kube-system` from auto-enroll
- but there are select workloads/pods with those excluded namespaces I do want to auto-enroll?

This is a relatively common scenario.

We can of course make auto-enroll a behavior that kicks in in the _absence of_ express user intent/labels which works well enough. But in this case it can't be an enforcement mechanism.

**Comment by bleggett** [comment]
Some good discussion out of the WG on this: https://docs.google.com/document/d/1SMlwliEnthgq7r2PjpLl1kCq3t8rAMbgu6r_lDAXJ0w/edit?tab=t.0#heading=h.yizcbnt4yh96

I think we know what we want to do here for autoenroll UX, but I'm going to look at solving #49009 in earnest first, since I do believe it to be a pretty significant blocker.

**Comment by Winbobob** [comment]
Not stale
